klippy/mathutil.py

```diff
@@ -135,3 +135,76 @@ def matrix_sub(m1, m2):
 
 def matrix_mul(m1, s):
     return [m1[0]*s, m1[1]*s, m1[2]*s]
+
+######################################################################
+# Polynomial Helper Classes and Functions
+######################################################################
+
+def calc_determinant(matrix):
+    m = matrix
+    aei = m[0][0] * m[1][1] * m[2][2]
+    bfg = m[1][0] * m[2][1] * m[0][2]
+    cdh = m[2][0] * m[0][1] * m[1][2]
+    ceg = m[2][0] * m[1][1] * m[0][2]
+    bdi = m[1][0] * m[0][1] * m[2][2]
+    afh = m[0][0] * m[2][1] * m[1][2]
+    return aei + bfg + cdh - ceg - bdi - afh
+
+class Polynomial2d:
+    def __init__(self, a, b, c):
+        self.a = a
+        self.b = b
+        self.c = c
+
+    def __call__(self, xval):
+        return self.c * xval * xval + self.b * xval + self.a
+
+    def get_coefs(self):
+        return (self.a, self.b, self.c)
+
+    def __str__(self):
+        return "%f, %f, %f" % (self.a, self.b, self.c)
```

Comment:
I don't think `Polynomial2d` is a good fit for mathutil.py.  We can certainly add the linear squares code and/or the matrix determinant code to mathutil.py.  (Interestingly, I think @dmbutyugin recently made a similar implementation for his generic_cartesian code.)

The `EddyDriftCompensation` code has some subtle dependencies on `Polynomianl2d` that I'm unsure would be apparent to developers looking at just mathutil.py (eg, the `str()` format becomes a user-facing api as it is written directly to the config file).

One simple solution would be to relocate `Polynomial2d` to temperature_probe.py .

Reply:
Sounds good.

Reply:
Yes, the code can be found here
https://github.com/dmbutyugin/klipper/blob/5f4df09c99d9e92a3f9896c9e8107659d0efc0d0/klippy/mathutil.py#L143-L152
and the calculation of position using pseudoinverse:
https://github.com/dmbutyugin/klipper/blob/generic-cartesian/klippy/kinematics/generic_cartesian.py#L124-L128
which is also an least squares problem solution. I think this one is a bit less verbose, so feel free to use, but up to you.